debug/fetch_data.c

```diff
@@ -28,53 +28,69 @@ static const char* getfield(char* line, int num)
 #define NCHANNELS 32
 #define NSESSIONS 8
 
+#define FEXIT(series, size) {*(series) = NULL; *(size) = NULL; return 0;}
+
+#define SFOR(i, j, limit_i, limit_j) \
+    for (size_t (i) = 0; (i) < (limit_i); ++(i)) \
+        for (size_t (j) = 0; (j) < (limit_j); ++(j))
+
 /*
  * Parses subject-specific files in train and 
  * returns the subrange corresponding to specified activity
  */
-static void fetch_activity_data(int pid, int chid, int evid, 
-        double **series, size_t *series_size) {
+static size_t fetch_activity_data(int pid, int chid, int evid, int sid,
+        double* **series, size_t* *series_size) {
     char buf[BUF_SIZE];
-    for (int sid = 0; sid < NSESSIONS; ++sid) {
-        // Parsing next session
-        snprintf(buf, BUF_SIZE, 
-                "data/train/subj%d_series%d_data.csv", pid + 1, sid + 1);
-        FILE *dataf = fopen(buf, "r");
-        snprintf(buf, BUF_SIZE, 
-                "data/train/subj%d_series%d_events.csv", pid + 1, sid + 1);
-        FILE *eventsf = fopen(buf, "r");
-        size_t n_values = 0;
-        series[sid] = (double *) malloc(sizeof(double));
-        // Eat-up both csv headers
-        fgets(buf, BUF_SIZE, dataf);
-        fgets(buf, BUF_SIZE, eventsf);
-        while (fgets(buf, BUF_SIZE, eventsf)) {
-            char *tmp = strdup(buf);
-            if (strcmp(getfield(tmp, evid+2), "0") == 0) {
-                // No activity on event of interest registered -> skip
-                free(tmp);
-                continue;
-            }
-            free(tmp);
-            fgets(buf, BUF_SIZE, dataf);
-            tmp = strdup(buf);
-            series[sid][n_values++] = 
-                strtod(getfield(tmp, chid+2), NULL);
-            series[sid] = (double *)
-                realloc(series[sid], (n_values + 1) * sizeof(double));
+    snprintf(buf, BUF_SIZE, 
+            "data/train/subj%d_series%d_data.csv", pid + 1, sid + 1);
+    FILE *dataf = fopen(buf, "r");
+    snprintf(buf, BUF_SIZE, 
+            "data/train/subj%d_series%d_events.csv", pid + 1, sid + 1);
+    FILE *eventsf = fopen(buf, "r");
+    size_t n_values = 0, n_trials = 0;
+    int new = 1;
+    *series = NULL;
+    *series_size = NULL;
+    // Eat-up both csv headers
+    if (fgets(buf, BUF_SIZE, dataf) == NULL) FEXIT(series, series_size);
+    if (fgets(buf, BUF_SIZE, eventsf) == NULL) FEXIT(series, series_size);
+    while (fgets(buf, BUF_SIZE, eventsf)) {
+        char *tmp = strdup(buf);
+        if (strcmp(getfield(tmp, evid+2), "0") == 0) {
+            // No activity on event of interest registered -> skip
+            new = 1;
             free(tmp);
+            continue;
```

Comment:
HUGE BUG up here -- I don't eat up corresponding line in dataf by continue'ing here. All results invalidated.